File: npmjs/pray-cli/src/distribution.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { PrayError } from "./errors.js";

export const DISTRIBUTION_CONFIG_SPEC = "pray-distribution-config-1";
const TORRENT_PROTOCOL = "torrent";
const DISTRIBUTION_CONFIG_FIELDS = new Set([
  "spec",
  "protocols",
  "bootstrap_trackers",
  "enable_dht",
]);

export interface RegistryDistributionSettings {
  spec: string;
  protocols: string[];
  bootstrapTrackers: string[];
  enableDht: boolean;
}

export function defaultDistributionSettings(): RegistryDistributionSettings {
  return {
    spec: DISTRIBUTION_CONFIG_SPEC,
    protocols: [],
    bootstrapTrackers: [],
    enableDht: false,
  };
}

export function allowsTorrent(settings: RegistryDistributionSettings): boolean {
  return settings.protocols.includes(TORRENT_PROTOCOL);
}

export function parseDistributionSettings(
  text: string,
): RegistryDistributionSettings {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(text) as Record<string, unknown>;
  } catch (error) {
    throw PrayError.parse(
      "distribution config",
      error instanceof Error ? error.message : String(error),
    );
  }
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    throw PrayError.parse("distribution config", "expected an object");
  }
  for (const key of Object.keys(data)) {
    if (!DISTRIBUTION_CONFIG_FIELDS.has(key)) {
      throw PrayError.parse("distribution config", `unsupported field: ${key}`);
    }
  }
  const settings: RegistryDistributionSettings = {
    spec: typeof data.spec === "string" ? data.spec : DISTRIBUTION_CONFIG_SPEC,
    protocols: Array.isArray(data.protocols)
      ? data.protocols.map((name) => String(name))
      : [],
    bootstrapTrackers: Array.isArray(data.bootstrap_trackers)
      ? data.bootstrap_trackers.map((name) => String(name))
      : [],
    enableDht: data.enable_dht === true,
  };
  validateDistributionSettings(settings);
  return settings;
}

export function readDistributionSettings(
  root: string,
): RegistryDistributionSettings {
  const path = join(root, "v1", "distribution.json");
  if (!existsSync(path)) {
    return defaultDistributionSettings();
  }
  return parseDistributionSettings(readFileSync(path, "utf8"));
}

export function writeDistributionSettings(
  root: string,
  settings: RegistryDistributionSettings = defaultDistributionSettings(),
): void {
  validateDistributionSettings(settings);
  const path = join(root, "v1", "distribution.json");
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(
    path,
    `${JSON.stringify(
      { spec: settings.spec, protocols: settings.protocols },
      null,
      2,
    )}\n`,
  );
}

function validateDistributionSettings(
  settings: RegistryDistributionSettings,
): void {
  if (settings.spec !== DISTRIBUTION_CONFIG_SPEC) {
    throw PrayError.parse(
      "distribution config",
      `unsupported distribution config spec: ${settings.spec}`,
    );
  }
  for (const name of settings.protocols) {
    if (name !== TORRENT_PROTOCOL) {
      throw PrayError.parse(
        "distribution config",
        `unsupported protocol: ${name}`,
      );
    }
  }
  if (settings.enableDht) {
    throw PrayError.parse(
      "distribution config",
      "DHT announce is not implemented",
    );
  }
  if (settings.bootstrapTrackers.length > 0 && !allowsTorrent(settings)) {
    throw PrayError.parse(
      "distribution config",
      "bootstrap_trackers requires protocols to include torrent",
    );
  }
}

File: rubygems/pray-cli/lib/pray/distribution.rb
# frozen_string_literal: true

require "json"
require "fileutils"

module Pray
  DISTRIBUTION_CONFIG_SPEC = "pray-distribution-config-1"
  TORRENT_PROTOCOL = "torrent"
  DISTRIBUTION_CONFIG_FIELDS = %w[spec protocols bootstrap_trackers enable_dht].freeze

  RegistryDistributionSettings = Struct.new(
    :spec, :protocols, :bootstrap_trackers, :enable_dht,
    keyword_init: true
  ) do
    def initialize(spec: DISTRIBUTION_CONFIG_SPEC, protocols: [], bootstrap_trackers: [], enable_dht: false)
      super
    end

    def allows_torrent?
      Array(protocols).include?(TORRENT_PROTOCOL)
    end
  end

  module Distribution
    module_function

    def default_settings
      RegistryDistributionSettings.new
    end

    def parse_settings(text)
      data = JSON.parse(text)
      unless data.is_a?(Hash)
        raise Error.parse("distribution config", "expected an object")
      end

      data.each_key do |key|
        next if DISTRIBUTION_CONFIG_FIELDS.include?(key)

        raise Error.parse("distribution config", "unsupported field: #{key}")
      end
      settings = RegistryDistributionSettings.new(
        spec: data.fetch("spec", DISTRIBUTION_CONFIG_SPEC),
        protocols: Array(data["protocols"]),
        bootstrap_trackers: Array(data["bootstrap_trackers"]),
        enable_dht: data["enable_dht"] == true
      )
      validate_settings!(settings)
      settings
    rescue JSON::ParserError => error
      raise Error.parse("distribution config", error.message)
    end

    def read_settings(root)
      path = File.join(root, "v1", "distribution.json")
      return default_settings unless File.file?(path)

      parse_settings(File.read(path))
    end

    def write_settings(root, settings = default_settings)
      validate_settings!(settings)
      path = File.join(root, "v1", "distribution.json")
      FileUtils.mkdir_p(File.dirname(path))
      payload = {"spec" => settings.spec, "protocols" => Array(settings.protocols)}
      File.write(path, JSON.pretty_generate(payload))
    end

    def validate_settings!(settings)
      unless settings.spec == DISTRIBUTION_CONFIG_SPEC
        raise Error.parse("distribution config", "unsupported distribution config spec: #{settings.spec}")
      end
      Array(settings.protocols).each do |name|
        next if name == TORRENT_PROTOCOL

        raise Error.parse("distribution config", "unsupported protocol: #{name}")
      end
      if settings.enable_dht
        raise Error.parse("distribution config", "DHT announce is not implemented")
      end
      return if Array(settings.bootstrap_trackers).empty? || settings.allows_torrent?

      raise Error.parse("distribution config", "bootstrap_trackers requires protocols to include torrent")
    end

    def fetch_settings(source_url)
      parse_settings(Registry.http_get(Registry.join_url(source_url, "v1/distribution.json")))
    rescue Error => error
      return default_settings if error.message.match?(/:\s*404\b/)

      raise
    end
  end
end
